text/0000-const-trait-impls.md

```diff
@@ -0,0 +1,1045 @@
+there is absolutely no difference (just like specifying `Fn() + FnOnce()` has a redundant `FnOnce()` bound).
+
+## Why do traits methods need to be marked as `(const)`
+
+
+
+## `(const) Destruct` super trait
+
+The `Destruct` marker trait is used to name the previously unnameable drop glue that every type has.
+It has no methods, as drop glue is handled entirely by the compiler,
+but in theory drop glue could become something one can explicitly call without having to resort to extracting the drop glue function pointer from a `dyn Trait`.
+
+Traits that have `self` (by ownership) methods, will almost always drop the `self` in these methods' bodies unless they are simple wrappers that just forward to the generic parameters' bounds.
+
+The following never drops `T`, because it's the job of `<T as Add>` to handle dropping the values.
+
+```rust
+struct NewType<T>(T);
+
+impl<T: (const) Add<Output = T>> const Add for NewType<T> {
+    type Output = Self;
+    fn add(self, other: Self) -> Self::Output {
+        NewType(self.0 + other.0)
+    }
+}
+```
+
+But if any code path could drop a value...
+
+```rust
+struct NewType<T>(T, bool);
+
+struct Error;
+
+impl<T: (const) Add<Output = T>> const Add for NewType<T> {
+    type Output = Result<Self, Error>;
+    fn add(self, other: Self) -> Self::Output {
+        if self.1 {
+            Ok(NewType(self.0 + other.0, other.1))
+        } else {
+            // Drops both `self.0` and `self.1`
+            Err(Error)
+        }
+    }
+}
+```
+
+... then we need to add a `(const) Destruct` bound to `T`, to ensure
+`NewType<T>` can be dropped.
+
+This bound in turn will be infectious to all generic users of `NewType` like
+
+```rust
+const fn add<T: (const) Add>(
+    a: NewType<T>,
+    b: NewType<T>,
+) -> Result<NewType<T::Output>, Error> {
+    a + b
+}
+```
+
+which now need a `T: (const) Destruct` bound, too.
+In practice we have noticed that a large portion of APIs will have a `(const) Destruct` bound.
+This bound has little value as an explicit bound that appears almost everywhere.
+Especially since it is a fairly straight forward assumption that a type that has trait impls with `const` methods will also have a `Drop::drop` method that is `const` or only contain `const Destruct` types.
+
+In the future we will also want to support `dyn (const) Trait` bounds, which invariably will require the type to implement `(const) Destruct` in order to fill in the function pointer for the `drop` slot in the vtable.
+While that can in generic contexts always be handled by adding more `(const) Destruct` bounds, it would be more similar to how normal `dyn` safety
+works if there were implicit `(const) Destruct` bounds for (most?) `(const) Trait` bounds.
+
+Thus we require that all `trait`s with `(const)` methods also have a `(const) Destruct` super trait bound to ensure users don't need to add `(const) Destruct` bounds everywhere.
+We may relax this requirement in the future or make it implied.
```

Comment:
Beautiful solution :)

Reply:
If the standard library adopts these bounds for its traits , it won’t be able to relax them in the future. Which likely means that most of the ecosystem won’t either.

Reply:
> Thus we require that all `trait`s with `(const)` methods also have a `(const) Destruct` super trait bound to ensure users don't need to add `(const) Destruct` bounds everywhere.

I know this is something I had proposed, but where @nikomatsakis and I ended up landing was pushing these bounds to the use-site, which would imply not requiring this bound.  This doesn't save space for making it implicit then, of course, but we could always still do that over an edition.